File: algo_library/rmq.cpp
// implementation of Fenwick/Binary Index Tree

#include <iostream>
#include <string>
#include <vector>
#include <algorithm>
#include <sstream>
#include <queue>
#include <deque>
#include <bitset>
#include <iterator>
#include <list>
#include <stack>
#include <map>
#include <unordered_map>
#include <set>
#include <unordered_set>
#include <functional>
#include <numeric>
#include <utility>
#include <limits>
#include <time.h>
#include <math.h>
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <assert.h>

using namespace std;


const int N = 10000;
const int K = 25;
int st[N][K+1];
int logs[N+1];

void build(int arr[]) {
  logs[1] = 0;
  for (int i = 2; i <= N; i++)
      logs[i] = logs[i/2] + 1;
  for (int i = 0; i < N; i++) {
    st[i][0] = arr[i];
  }
  for (int j = 1; j <= K; j++)
    for (int i = 0; i + (1 << j) <= N; i++)
        st[i][j] = min(st[i][j-1], st[i + (1 << (j - 1))][j - 1]);
}

int rmquery(int L, int R) {
  int j = floor(log2(R - L + 1));
  int minimum = min(st[L][j], st[R - (1 << j) + 1][j]);
  return minimum;
}


// test range min queries
int main() {
    int arr[N];
    map<pair<int,int>, int> rmins;
    for (int i = 0; i < 10; i++) {
        arr[i] = i+1;
    }
    for (int i = 0; i < 10; i++) {
        for (int j = i; j < 10; j++) {
            int rmin = *min_element(arr+i, arr+j);
            rmins[make_pair(i, j)] = rmin;
        }
    }

    build(arr);

    for (int i = 0; i < 10; i++) {
        for (int j = i; j < 10; j++) {
            if (rmquery(i, j) != rmins[make_pair(i,j)]) { cout << "ERR" << endl;}
        }
    }
}

File: algo_library/rmq.py
"""
Sparse table implementation
"""
import math
N = 100
K = int(math.ceil(math.log(N+1,2)))
RMQ = [[0 for _ in range(K)] for _ in range(N)]


def build(arr):
    for i in range(N):
        RMQ[i][0] = arr[i]
    for k in range(1, K):
        for i in range(0, N-2**k+1):
            RMQ[i][k] = min(RMQ[i][k-1], RMQ[i+2**(k-1)][k-1])

# inclusive, exclusive
def rmquery(i, j):
    length = j-i
    if length == 0: return None
    k = int(math.floor(math.log(length,2)))
    left = RMQ[i][k]
    right = RMQ[j-2**k][k]
    return min(left, right)


# TEST
"""
import random
seed = random.randint(0, 100)
print("seed", seed)
random.seed(seed)
# arr = list(range(1, N+1))
arr = [random.randint(0, 100) for _ in range(N)]

build(arr)

rmins = dict()
for i in range(N):
    for j in range(i, N):
        rmin = min(arr[i:j+1])
        rmins[(i,j+1)] = rmin

for i in range(N):
    for j in range(i+1, N+1):
        if rmquery(i,j) != rmins[(i,j)]:
                print(i, j, rmquery(i,j), rmins[(i,j)])
"""

# https://judge.yosupo.jp/problem/staticrmq
import math
_, Q = input().split()
arr = list(map(int, input().split()))
N = len(arr)
K = int(math.ceil(math.log(N+1,2)))
RMQ = [[0 for _ in range(K)] for _ in range(N)]
build(arr)
for _ in range(int(Q)):
    l, r = map(int, input().split())
    print(rmquery(l,r))
